### PR TITLE
[SYCL] Don't include <complex> from <sycl/sycl.hpp>

### PR DESCRIPTION
**Pull request by aelovikov-intel**
This addresses a part of a bigger issue of polluting the global namespace with math functions by implicitly including `<cmath>` from sycl headers. Internally, `<complex>` includes `<cmath>` so we shouldn't be including it as well, which this patch implements.

I'm doing it by providing our own version of <complex> that does `#include_next <complex>` and also provides needed functions/template specializations to support `std::complex`-related functionality whenever user program includes <complex> on its own. System include path for this is only provided in SYCL mode, so non-SYCL compilation flow is unaffected.

Unfortunately, MSVC doesn't have have `#include_next` so we use a hack to emulate it - via `#include <../include/complex>`.

### CODE REVIEW

File: sycl/test-e2e/GroupAlgorithm/inclusive_scan_sycl2020.cpp
Comment (bader) [inline sycl/test-e2e/GroupAlgorithm/inclusive_scan_sycl2020.cpp:8]: Please, add a test where `<sycl/sycl.hpp>` is included before `<complex>`.

File: clang/lib/Driver/ToolChains/SYCL.cpp
Comment (mdtoguchi) [inline clang/lib/Driver/ToolChains/SYCL.cpp:1140]: Could you add a change to https://github.com/intel/llvm/blob/d6e66a7ed77f5ac5a5a3775f35778dfb6e1b5813/clang/test/Driver/sycl-offload.c#L622 to cover expected header directories pulled in?

File: clang/test/Driver/sycl-offload.c
Comment (aelovikov-intel) [inline clang/test/Driver/sycl-offload.c:627]: Previous version didn't catch cases where we had extra includes between the ones we match (due to `{{.*}}bin` greedy regex). This update is both catching that and doesn't require extra long lines making it more readable.

File: clang/lib/Driver/ToolChains/SYCL.cpp
Comment (aelovikov-intel) [reply clang/lib/Driver/ToolChains/SYCL.cpp:1140]: Done.

**Review by mdtoguchi — APPROVED** [review]
OK for driver

File: clang/lib/Driver/ToolChains/SYCL.cpp
Comment (bader) [inline clang/lib/Driver/ToolChains/SYCL.cpp:1131]: Couple of style suggestions:
1. Update the comment at line 1125.
2. Please, re-arrange the code:
from 
```c++
// define path1
// define path2
// set path1 via CC1Args
// set path2 via CC1Args
```
to 
```c++
// define path1
// set path1 via CC1Args
// define path2
// set path2 via CC1Args
```

Comment (bader) [reply clang/lib/Driver/ToolChains/SYCL.cpp:1131]: After reading the code more carefully, I realized that suggestion 2 will require code duplication as we define all paths before we can set the first path due to order restrictions. If so, I suggest ignoring my second suggestion.

File: clang/lib/Driver/ToolChains/SYCL.cpp
Comment (bader) [inline clang/lib/Driver/ToolChains/SYCL.cpp:1129]:
```suggestion
  llvm::sys::path::append(P, "include");
  llvm::sys::path::remove_dots(P, /*remove_dot_dot=*/true);
  SmallString<128> SYCLP(P);
```

File: clang/lib/Driver/ToolChains/SYCL.cpp
Comment (bader) [inline clang/lib/Driver/ToolChains/SYCL.cpp:1134]:
```suggestion
  SmallString<128> STLWrappersPath(SYCLP);
```

I would also rename `P` to `IncludePath` and `SYCLP` to `SYCLPath`.

File: sycl/include/sycl/stl_wrappers/complex
Comment (keryell) [inline sycl/include/sycl/stl_wrappers/complex:3]: Long time since I have not seen this in use.
Perhaps a good opportunity to add some comments before explaining the basic design ideas?

File: clang/lib/Driver/ToolChains/SYCL.cpp
Comment (aelovikov-intel) [reply clang/lib/Driver/ToolChains/SYCL.cpp]: I'll leave this to a separate PR.

File: sycl/include/sycl/stl_wrappers/complex
Comment (aelovikov-intel) [reply sycl/include/sycl/stl_wrappers/complex:20]: Done, I think.

**Comment by aelovikov-intel** [comment]
@cperkinsintel , @intel/llvm-reviewers-runtime ping.

File: sycl/include/sycl/stl_wrappers/complex
Comment (cperkinsintel) [reply sycl/include/sycl/stl_wrappers/complex:20]: Is `#include_next` supported by MSVC?  A cursory google search says it is not.

**Comment by cperkinsintel** [comment]
So if a user wants to use complex numbers, they need to `#include <complex>` themselves, and do so before `#include <sycl/sycl.hpp>`, correct? Or am I misunderstanding how this works?

**Comment by aelovikov-intel** [comment]
> So if a user wants to use complex numbers, they need to `#include <complex>` themselves, and do so before `#include <sycl/sycl.hpp>`, correct? Or am I misunderstanding how this works?

Both before/after would work - see https://github.com/intel/llvm/pull/11196/files#diff-d8871abe2a3a33ba61392bd95ccf6f1caab18118964cd0814868b73afeae4af0.

**Review by cperkinsintel — APPROVED** [review]
ship it

**Comment by aelovikov-intel** [comment]
@mdtoguchi , there was small driver change after your approval to support `-fsycl-host-compiler`. Can you take a look post-commit, please?

**Pull request merged**
